### PR TITLE
God Class Apply Refactoring

### PR DESCRIPTION
**Pull request by 123a6bcw**

### CODE REVIEW

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/refactoring/ExtractClassRefactoring.java [hunk 1]
```diff
@@ -0,0 +1,3071 @@
...
+                //if an assigned field is not found in left hand side, then replace all accessed fields in left hand side
+                int j = 0;
+                List<PsiExpression> newAccessedVariables = expressionExtractor.getVariableInstructions(newLeftHandSide);
+                for (PsiExpression expression2 : newAccessedVariables) {
+                    PsiReferenceExpression oldAccessedVariableReferenceExpression = (PsiReferenceExpression) expression2;
+                    PsiVariable oldAccessedVariable = (PsiVariable) oldAccessedVariableReferenceExpression.resolve();
+
+                    PsiReferenceExpression newAccessedVariableReferenceExpression = (PsiReferenceExpression) newAccessedVariables.get(j);
+                    if (oldAccessedVariable != null) {
+                        if (oldAccessedVariable instanceof PsiField && !oldAccessedVariable.getModifierList().hasModifierProperty(Ps
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/refactoring/ExtractClassRefactoring.java]: I think we should remove unused code

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassPanel.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import com.intellij.openapi.application.TransactionGuard;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.research.intellijdeodorant.ide.ui.abstractrefactorings.ExtractClassRefactoringType;
+import org.jetbrains.research.intellijdeodorant.ide.ui.abstractrefactorings.ExtractClassRefactoringType.AbstractExtractClassRefactoring;
+import org.jetbrains.research.intellijdeodorant.ide.ui.abstractrefactorings.RefactoringType.AbstractCandidateRefactoring;
+
+import java.util.Collections;
+
+/**
+ * Panel for Type-State Checking refactorings.
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassPanel.java]: Please, fix the comment

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+import javax.swing.*;
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import java.awt.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+
+public class GodClassUserInputDialog extends RefactoringDialog {
+    private static final String TITLE = "Extracted Class Name";
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Move to IntelliJDeodorantBundle

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    @Nullable
+    @Override
+    protected JComponent createCenterPanel() {
+        return new JPanel();
+    }
+
+    private void initialiseControls() {
+        extractedClassNameLabel.setText("Extracted Class Name");
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Move to IntelliJDeodorantBundle

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+                handleInputChanged(extractedClassNameField);
+            }
+
+            @Override
+            public void changedUpdate(DocumentEvent e) {
+                handleInputChanged(extractedClassNameField);
+            }
+        });
+
+        restoreButton.setText("Restore Defaults");
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Move to IntelliJDeodorantBundle

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+        delegateButton.setText("Keep original public methods as delegates to the extracted methods");
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Move to IntelliJDeodorantBundle

File: src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ClusterSizeComparator.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package org.jetbrains.research.intellijdeodorant.core.distance;
+
+import java.util.Comparator;
+
+public class ClusterSizeComparator implements Comparator<ExtractClassCandidateRefactoring> {
+
+	public int compare(ExtractClassCandidateRefactoring o1,
+			ExtractClassCandidateRefactoring o2) {
+		if(o1.getExtractedEntities().size() > o2.getExtractedEntities().size()) {
```
[the remark this answers is not in the document]
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ClusterSizeComparator.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ExtractClassCandidateRefactoring.java [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    private boolean containsFieldAccessOfEnclosingClass(MyMethod method) {
+        if (method.getMethodObject().containsFieldAccessOfEnclosingClass()) {
+            return true;
+        } else
+            return false;
+    }
+
+    private boolean overridesMethod(MyMethod method) {
+        if (method.getMethodObject().overridesMethod()) {
+            //System.out.println(this.toString() + "\toverrides method of superclass");
```
[the remark this answers is not in the document]
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ExtractClassCandidateRefactoring.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ExtractClassCandidateRefactoring.java [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+            return true;
+        } else
+            return false;
+    }
+
+    private boolean containsSuperMethodInvocation(MyMethod method) {
+        if (method.getMethodObject().containsSuperMethodInvocation()) {
+            //System.out.println(this.toString() + "\tcontains super method invocation");
```
[the remark this answers is not in the document]
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ExtractClassCandidateRefactoring.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ExtractClassCandidateRefactoring.java [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+            return true;
+        } else
+            return false;
+    }
+
+    private boolean isSynchronized(MyMethod method) {
+        if (method.getMethodObject().isSynchronized()) {
+            //System.out.println(this.toString() + "\tis synchronized");
```
[the remark this answers is not in the document]
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/core/distance/ExtractClassCandidateRefactoring.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/core/ast/util/ExpressionExtractor.java [hunk 11]
```diff
@@ -444,6 +494,8 @@
             if (instanceChecker.instanceOf(arrayInitializer))
                 expressionList.add(arrayInitializer);
         } else if (expression instanceof PsiReferenceExpression) {
+            //TODO plugin does not find field accesses such as FIELD.toString()
```
[the remark this answers is not in the document]
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/core/ast/util/ExpressionExtractor.java]: Removed

File: src/main/java/org/jetbrains/research/intellijdeodorant/core/ast/MethodObject.java [hunk 12]
```diff
@@ -709,4 +711,43 @@ public String toString() {
...
+
+            for (PsiElement thisExpressionElement : psiElements) {
+                PsiThisExpression thisExpression = (PsiThisExpression) thisExpressionElement;
+                if (thisExpression.getQualifier() != null) {
+                    return true;
+                }
+            }
+        }
+
+        /*
```
[the remark this answers is not in the document]
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/core/ast/MethodObject.java]: Removed

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/refactoring/ExtractClassRefactoring.java (hunk 1 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/refactoring/ExtractClassRefactoring.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassPanel.java (hunk 2 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassPanel.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java (hunk 3 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java (hunk 4 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java (hunk 5 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java (hunk 6 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Done

File: src/main/java/org/jetbrains/research/intellijdeodorant/util/math/Edge.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+
+        if (o instanceof Edge) {
+            Edge edge = (Edge) o;
+            return this.from.equals(edge.from) && this.to.equals(edge.to);
+        }
+        return false;
+        // :thi
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/util/math/Edge.java]: ?

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+        }
+
+        mayApplyRefactoring(true);
+        setMessage("");
+    }
+
+    @Override
+    protected void doAction() {
+        if (isPreviewUsages()) {
+            //TODO
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: ?

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+        JComponent emptyComponent = new JComponent() {
+            @Override
+            public void setInheritsPopupMenu(boolean value) {
+                super.setInheritsPopupMenu(value);
+            }
+        };
+
+        builder.addLabeledComponent(restoreButton, emptyComponent);
+
+        mainPanel = builder.addVerticalGap(5).getPanel();
```
Comment (onewhl) [inline src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: What is a magic number?

File: src/main/java/org/jetbrains/research/intellijdeodorant/util/math/Edge.java (hunk 13 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/util/math/Edge.java]: I honestly have no idea how and why this comment appeared here. Removed.

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java (hunk 14 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Removed (for now)

File: src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java (hunk 15 above)
Comment (123a6bcw) [reply src/main/java/org/jetbrains/research/intellijdeodorant/ide/ui/GodClassUserInputDialog.java]: Magic straight from the IDEA source (everywhere I looked they just write `5`). Well, ok, extracted static final field.

**Comment by 123a6bcw** [comment]
Latest commit fixes issue fixed in the original plugin
https://github.com/tsantalis/JDeodorant/issues/76
https://github.com/tsantalis/JDeodorant/commit/251a14ee1490cf0083b408d61d413aa3ac64a143

**Pull request merged**